README.md

```diff
@@ -157,206 +235,187 @@ follow up proposal.
+  run<R>(value: T, callback: () => R): R;
 
-As the value of AsyncLocal has to be fetched from its own store, i.e. the
-AsyncLocal object. From arbitrary execution context in an async execution
-flow, users have to declare their own AsyncLocal to get their value
-propagates along with an async execution flow.
+  get(): T;
+}
+```
 
-AsyncLocal propagates values along the logical async execution flow.
+`AsyncContext.p.run()` and `AsyncContext.p.get()` sets and gets the current
+value of an async execution flow. `AsyncContext.wrap()` allows you to opaquely
+capture the current value of all `AsyncContexts` and execute the callback at a
+later time with as if those values were still the current values (a snapshot and
+restore).
 
-```js
-const asyncLocal = new AsyncLocal();
+```typescript
+const context = new AynsContext();
 
-(function main() {
-  asyncLocal.setValue('main');
+// Sets the current value to 'top', and executes the `main` function.
+context.run('top', main);
 
+function main() {
+  // Context is maintained through other platform queueing.
   setTimeout(() => {
-    console.log(asyncLocal.getValue()); // => 'main'
-    asyncLocal.setValue('first timer');
-    setTimeout(() => {
-      console.log(asyncLocal.getValue()); // => 'first timer'
-    }, 1000);
-  }, 1000);
+    console.log(context.get()); // => 'top'
 
-  setTimeout(() => {
-    console.log(asyncLocal.getValue()); // => 'main'
-    asyncLocal.setValue('second timer');
-    setTimeout(() => {
-      console.log(asyncLocal.getValue()); // => 'second timer'
-    }, 1000);
-  }, 1000);
-})();
-```
+    context.run('A', () => {
+      console.log(context.get()); // => 'A'
 
-> Note: There are controversial thought on the dynamic scoping and AsyncLocal,
-> checkout [SCOPING.md][] for more details.
+      setTimeout(() => {
+        console.log(context.get()); // => 'A'
+      }, randomTimeout());
+    });
+  }, randomTimeout());
+
+  // Context runs can be nested.
+  context.run('B', () => {
+    console.log(context.get()); // => 'B'
 
-The optional `valueChangedListener` will be called each time the value in the
-current async flow has been updated by explicit `AsyncLocal.setValue` call. It
-can be treated as a property setter of an object.
+    setTimeout(() => {
+      console.log(context.get()); // => 'B'
+    }, randomTimeout());
+  });
 
-The motivation for the `valueChangedListener` is that we can have a cleaner way
-to monitor the value changes of an AsyncLocal without any `setValue` wrappers.
+  // Context was restored after the previous run.
+  console.log(context.get()); // => 'top'
 
-As the `valueChangedListener` is only going to be trigger by explicit value
-set, codes where can trigger the `valueChangedListener` is strictly restricted
-since the code where trigger the listener has to explicitly refer to the
-instance of AsyncLocal.
+  // Maintains the state of all AsyncContext's at this moment.
```

Comment:
I'd find the term "capture" used in the API introduction can be more intuitive here.

```suggestion
  // Captures the state of all AsyncContext's at this moment.
```